### PR TITLE
Client judge

### PR DESCRIPTION
**Pull request by rogercloud**

### CODE REVIEW

File: curp/src/client.rs [hunk 1]
```diff
@@ -1 +1,139 @@
...
+}
+
+impl<C> Client<C>
+where
+    C: Command + 'static,
+{
+    /// Create a new protocol client based on the addresses
+    #[inline]
+    pub async fn new(leader: &SocketAddr, addrs: &[SocketAddr]) -> Option<Self> {
+        let ep = Endpoint::bind("127.0.0.1:0")
```
Comment (pzheng1025) [inline curp/src/client.rs:48]: Will this endpoint use random port number?

File: curp/src/client.rs [hunk 2]
```diff
@@ -1 +1,139 @@
...
+
+        let mut ok_cnt: usize = 0;
+        let mut max_term = 0;
+        let mut execute_result: Option<C::ER> = None;
+
+        while ok_cnt
+            >= (ft
+                .wrapping_add(ft.wrapping_add(1).wrapping_div(2))
+                .wrapping_add(1))
+            && execute_result.is_some()
```
Comment (pzheng1025) [inline curp/src/client.rs:85]: Is this logic reversed?

Comment (pzheng1025) [reply curp/src/client.rs:85]: What is the reason of using wrapping_add instead of overflowing_add?

File: curp/src/server.rs [hunk 3]
```diff
@@ -1,39 +1,129 @@
...
+    /// handle Propose
+    #[rpc]
+    async fn propose(&self, p: Propose<C>) -> ProposeResponse<C> {
+        self.inner.propose(p).await
+    }
+
+    /// Run a new rpc server
+    #[allow(dead_code)]
+    async fn run(is_leader: bool, term: u64, others: &[SocketAddr]) -> Result<(), ServerError> {
+        let rx_ep = Endpoint::bind(format!("0.0.0.0:{DEFAULT_SERVER_PORT}")).await?;
```
Comment (pzheng1025) [inline curp/src/server.rs:36]: servers use same port then we cannot test on same machine.

File: curp/src/cmd.rs [hunk 4]
```diff
@@ -19,6 +19,9 @@ pub trait Command:
     /// Get keys of the command
     fn keys(&self) -> &[Self::K];
 
+    /// Get cmd id
+    fn id(&self) -> &CmdId;
```
Comment (pzheng1025) [inline curp/src/cmd.rs:23]: Is this something like propose id?

File: curp/src/client.rs (hunk 1 above)
Comment (rogercloud) [reply curp/src/client.rs:48]: Yes, port 0 means pick a random port number, it's the client end.

File: curp/src/client.rs (hunk 2 above)
Comment (rogercloud) [reply curp/src/client.rs:85]: Yes, good catch. we can make sure ft the result is less than usize limit for that `ft = peer_cnt / 2`.

File: curp/src/cmd.rs (hunk 4 above)
Comment (rogercloud) [reply curp/src/cmd.rs:23]: It's propose id actually, let me rename it.

File: curp/src/server.rs (hunk 3 above)
Comment (rogercloud) [reply curp/src/server.rs:36]: Let me add a customized port option.

**Review by pzheng1025 — APPROVED** [review]

**Pull request merged**